File: packages/shared-rules-engine/src/unlocks.ts
import type { NowLocal, TempUnlock, TempUnlockScope } from "./types.js";

export interface UnlockEffects {
  bypassSchedule: boolean;
  bypassAllQuotas: boolean;
  extraTotalMinutes: number;
  extraPerTargetMinutes: Record<string, number>;
}

export function activeUnlocks(
  unlocks: TempUnlock[],
  now: NowLocal,
): TempUnlock[] {
  return unlocks.filter(
    (u) => !u.revoked && u.expiresAtMs > now.epochMs,
  );
}

export function computeUnlockEffects(
  unlocks: TempUnlock[],
  now: NowLocal,
): UnlockEffects {
  const active = activeUnlocks(unlocks, now);
  const effects: UnlockEffects = {
    bypassSchedule: false,
    bypassAllQuotas: false,
    extraTotalMinutes: 0,
    extraPerTargetMinutes: {},
  };

  for (const u of active) {
    if (u.scope === "schedule") {
      effects.bypassSchedule = true;
    }
    if (u.scope === "schedule+quotas") {
      effects.bypassSchedule = true;
      effects.bypassAllQuotas = true;
    }
    if (u.scope === "add-minutes" && u.additionalMinutes) {
      if (u.target === "total") {
        effects.extraTotalMinutes += u.additionalMinutes;
      } else if (u.target) {
        effects.extraPerTargetMinutes[u.target] =
          (effects.extraPerTargetMinutes[u.target] ?? 0) +
          u.additionalMinutes;
      }
    }
  }

  return effects;
}

export function mostPermissiveScope(
  scopes: TempUnlockScope[],
): TempUnlockScope | null {
  if (scopes.includes("schedule+quotas")) return "schedule+quotas";
  if (scopes.includes("schedule")) return "schedule";
  if (scopes.includes("add-minutes")) return "add-minutes";
  return null;
}

File: apps/windows-agent/src/screen_time_rules/unlocks.py
from __future__ import annotations

from dataclasses import dataclass, field

from screen_time_rules.models import NowLocal, TempUnlock


@dataclass
class UnlockEffects:
    bypass_schedule: bool = False
    bypass_all_quotas: bool = False
    extra_total_minutes: float = 0
    extra_per_target_minutes: dict[str, float] = field(default_factory=dict)


def active_unlocks(unlocks: list[TempUnlock], now: NowLocal) -> list[TempUnlock]:
    return [u for u in unlocks if not u.revoked and u.expires_at_ms > now.epoch_ms]


def compute_unlock_effects(unlocks: list[TempUnlock], now: NowLocal) -> UnlockEffects:
    effects = UnlockEffects()
    for u in active_unlocks(unlocks, now):
        if u.scope == "schedule":
            effects.bypass_schedule = True
        if u.scope == "schedule+quotas":
            effects.bypass_schedule = True
            effects.bypass_all_quotas = True
        if u.scope == "add-minutes" and u.additional_minutes:
            if u.target == "total":
                effects.extra_total_minutes += u.additional_minutes
            elif u.target:
                effects.extra_per_target_minutes[u.target] = (
                    effects.extra_per_target_minutes.get(u.target, 0) + u.additional_minutes
                )
    return effects
